sale_order_mass_action/wizards/sale_order_mass_action_wizard.py

```diff
@@ -4,22 +4,50 @@
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
 from odoo import api, fields, models
+from odoo.osv import expression
 
 
 class SaleOrderMassActionWizard(models.TransientModel):
 
     _name = "sale.order.mass.action.wizard"
     _description = "Sale Order Mass Action"
 
-    confirm = fields.Boolean(
-        help="Check this box if you want to confirm all the selected quotations."
+    action = fields.Selection(
+        [
+            ("confirm", "Confirm"),
+            ("quotation_sent", "Send Quotation"),
+            ("done", "Lock"),
+            ("cancel", "Cancel"),
+            ("draft", "Set to Draft"),
+        ]
     )
 
-    def _get_sale_order_confirm_domain(self):
-        return [
-            ("id", "in", self.env.context.get("active_ids")),
-            ("state", "in", ("draft", "sent")),
-        ]
+    def _get_sale_order_state_domain(self):
+        self.ensure_one()
+        if self.action == "confirm":
+            return [("state", "in", ("draft", "sent"))]
+        elif self.action == "quotation_sent":
+            return [("state", "=", "draft")]
+        elif self.action == "done":
+            return [("state", "=", "sale")]
+        elif self.action == "cancel":
+            return [("state", "in", ("cancel", "draft"))]
+        elif self.action == "draft":
+            return [("state", "in", ("cancel", "sent"))]
+
+    def _get_sale_order_domain(self):
+        self.ensure_one()
+        domain = [("id", "in", self.env.context.get("active_ids"))]
+        domain = expression.AND([domain, self._get_sale_order_state_domain()])
+        return domain
+
+    def _launch_action(self, sale_orders):
+        self.ensure_one()
+        action = "action_" + self.action
+        if hasattr(sale_orders, action):
+            getattr(sale_orders, action)()
+            return True
+        return False
```

Comment:
minor question (not blocking) : if the action is not found in the hasattr, I expect the function raise an error. don't you think ?

Reply:
Maybe. In this case, the user won't have the result notification with the updated sale orders.

But we could improve it and display an error notify in that case

Reply:
@legalsylvain What do you think ?

Reply:
But in which case the action doesn't exist ? Sorry but i dont understand that test.